src/Plutip/UtxoDistribution.purs

```diff
@@ -0,0 +1,175 @@
+module Plutip.UtxoDistribution
+  ( class UtxoDistribution
+  , decodeWallets
+  , keyWallets
+  , encodeDistribution
+  , transferFundsFromEnterpriseToBase
+  , withStakeKey
+  ) where
+
+import Prelude
+
+import Contract.Address
+  ( PaymentPubKeyHash
+  , StakePubKeyHash
+  , getNetworkId
+  , getWalletAddress
+  , ownPaymentPubKeyHash
+  , ownStakePubKeyHash
+  , payPubKeyHashEnterpriseAddress
+  )
+import Contract.Monad (Contract, liftContractM, liftedE, liftedM)
+import Contract.Prelude (foldM, foldMap, null)
+import Contract.ScriptLookups as Lookups
+import Contract.Transaction
+  ( TransactionOutput(TransactionOutput)
+  , awaitTxConfirmed
+  , balanceAndSignTxE
+  , signTransaction
+  , submit
+  )
+import Contract.TxConstraints as Constraints
+import Contract.Utxos (utxosAt)
+import Contract.Wallet (withKeyWallet)
+import Data.Array as Array
+import Data.FoldableWithIndex (foldMapWithIndex)
+import Data.List (List, (:))
+import Data.Maybe (Maybe(Nothing, Just))
+import Data.Newtype (unwrap, wrap)
+import Data.Tuple (Tuple(Tuple))
+import Data.Tuple.Nested (type (/\), (/\))
+import Plutip.Types
+  ( InitialUTxO
+  , InitialUTxOWithStakeKey(InitialUTxOWithStakeKey)
+  , PrivateKeyResponse(PrivateKeyResponse)
+  , UtxoAmount
+  )
+import Plutus.Types.Transaction (Utxo)
+import Type.Prelude (Proxy(Proxy))
+import Wallet.Key
+  ( KeyWallet
+  , PrivatePaymentKey(PrivatePaymentKey)
+  , PrivateStakeKey
+  , privateKeysToKeyWallet
+  )
+
+-- | A type class that implements a type-safe interface for specifying UTXO
+-- | distribution for wallets.
+-- | Number of wallets in distribution specification matches the number of
+-- | wallets provided to the user.
+class UtxoDistribution distr wallets | distr -> wallets where
+  encodeDistribution :: distr -> Array (Array UtxoAmount)
+  decodeWallets :: distr -> Array PrivateKeyResponse -> Maybe wallets
+  keyWallets :: Proxy distr -> wallets -> Array KeyWallet
+
+instance UtxoDistribution Unit Unit where
+  encodeDistribution _ = []
```

Comment:
@klntsky Should this line be this instead:
```suggestion
  encodeDistribution _ = [[]]
```
Currently when the `Unit` instance is used it can break things because a `PrivateKeyResponse` is not returned for `unit`. For example `distr = unit /\ aliceUtxos` will only request one private key, and this messes with the wallet decoding. The change above fixes that though

Reply:
I'm just unsure what the purpose of the unit instance was. I assume it was to represent an empty wallet

Reply:
Actually no that doesn't make sense because the wallet is also `Unit`

Reply:
The idea is to not ask for any wallet and not use it, in cases when we are just using queries.

Reply:
Would it be okay to make this change and just ignore the returned private key? Because currently we're able to keep the recursion simple by assuming there will be a private key for each distribution, but the assumption is violated by this instance. I think the alternatives would be quite messy

Reply:
I don't see a problem with it, the wallet will just be `unit`.

Reply:
Ended up going with a different solution the fixed this as well as another bug, see https://github.com/Plutonomicon/cardano-transaction-lib/pull/838/commits/bea82b5772e432d6847eb9a95640c107854cd38c